polaris-service/src/main/java/org/apache/polaris/service/task/TableCleanupTaskHandler.java

```diff
@@ -158,6 +159,13 @@ public boolean handleTask(TaskEntity cleanupTask) {
         for (PolarisBaseEntity createdTask : createdTasks) {
           taskExecutor.addTaskHandlerContext(createdTask.getId(), CallContext.getCurrentContext());
         }
+
+        tableMetadata.previousFiles().stream()
+            .map(TableMetadata.MetadataLogEntry::file)
+            .forEach(fileIO::deleteFile);
+        tableMetadata.statisticsFiles().stream()
+            .map(StatisticsFile::path)
+            .forEach(fileIO::deleteFile);
```

Comment:
We should either add retries for these or submit them as separate tasks. As is, if one of these files fails to delete, we'll retry the whole task and resubmit a new task for each manifest. If the manifests are already deleted when we retry, we'll get stuck in a retry loop.

Reply:
Thanks for the suggestion, Michael! I could refactor the process by deleting these files in a new task and executing them separately, similar to how we handle manifest file deletions. Does that align with what you recommended approach?

Reply:
Sorry, do you mean adding retry within the `TableCleanupTaskHandler.java`? If so, that's fine with me. I'd at least use an Executor to attempt the execution in parallel (i.e., so file2 isn't blocked during the delay between retries for file1).

TBH, I'm not very familiar with how Iceberg generates the statistic files. Does there tend to be one per snapshot? one per data file? If the latter, we could be talking about a very large number of files. If that's the case, I think submitting separate tasks with batches of files makes sense.

Reply:
They are per snapshots, https://iceberg.apache.org/spec/#table-statistics.

Reply:
Is there only one statistics file per snapshot? The spec is not clear:
> A table can contain many statistics files associated with different table snapshots.

Unlike the partition statistics file, which is very clear:
> Each table snapshot may be associated with at most one partition statistics file.

Reply:
Could be multiple statistics files as the following code shows:

https://github.com/apache/iceberg/blob/5439cbdb278232779fdd9a392bbf57f007f9bda0/core/src/main/java/org/apache/iceberg/TableMetadata.java#L535-L535

Reply:
> They are per snapshots, https://iceberg.apache.org/spec/#table-statistics.

> Could be multiple statistics files as the following code shows:
> 
> https://github.com/apache/iceberg/blob/5439cbdb278232779fdd9a392bbf57f007f9bda0/core/src/main/java/org/apache/iceberg/TableMetadata.java#L535-L535

Thank you for providing the context, @flyrain :)

> Sorry, do you mean adding retry within the `TableCleanupTaskHandler.java`? If so, that's fine with me. I'd at least use an Executor to attempt the execution in parallel (i.e., so file2 isn't blocked during the delay between retries for file1).
> 
> TBH, I'm not very familiar with how Iceberg generates the statistic files. Does there tend to be one per snapshot? one per data file? If the latter, we could be talking about a very large number of files. If that's the case, I think submitting separate tasks with batches of files makes sense.

@collado-mike Yes, if we go for retries, the logic will be within `TableCleanupTaskHandler.java`. But considering the information provided by Yufei, maybe separate tasks will be a more appropriate approach for stats file deletion?